app/gui2/src/components/ComponentBrowser/__tests__/input.test.ts

```diff
@@ -264,3 +289,96 @@ test.each([
     })
   },
 )
+
+interface ImportsCase {
+  description: string
+  suggestionId: number
+  existingImports: { import: Import; span: ContentRange }[]
+  initialCode?: string
+  manuallyEditedCode?: string
+  expectedCode: string
+  expectedImports: RequiredImport[]
+}
+
+test.each([
+  {
+    description: 'Basic case of adding required import',
+    suggestionId: 3,
+    existingImports: [],
+    expectedCode: 'Table.new ',
+    expectedImports: [
+      {
+        kind: 'Unqualified',
+        from: unwrap(tryQualifiedName('Standard.Base')),
+        import: unwrap(tryIdentifier('Table')),
+      },
+    ],
+  },
+  {
+    description: 'No need to add already existing import',
+    suggestionId: 3,
+    existingImports: [
+      {
+        import: {
+          from: unwrap(tryQualifiedName('Standard.Base')),
+          imported: { kind: 'All', except: [] },
+        },
+        span: [0, 0],
+      },
+    ],
+    expectedCode: 'Table.new ',
+    expectedImports: [],
+  },
```

Comment:
I'm not sure if we should check that _during_ node editing. We could just keep the import and check for already present imports once committed.

Reply:
We only do it after editing is accepted, but from an architectural standpoint, you’re right. I moved part of the logic to `graph` store.